upup/pkg/kutil/delete_cluster.go

```diff
@@ -1401,6 +1423,102 @@ func FindAutoScalingLaunchConfigurations(cloud fi.Cloud, securityGroups sets.Str
 	return trackers, nil
 }
 
+func FindNatGateways(cloud fi.Cloud, routeTableIds sets.String) ([]*ResourceTracker, error) {
+	if len(routeTableIds) == 0 {
+		return nil, nil
+	}
+
+	c := cloud.(awsup.AWSCloud)
+
+	natGatewayIds := sets.NewString()
+
+	{
+		request := &ec2.DescribeRouteTablesInput{}
+		for routeTableId := range routeTableIds {
+			request.RouteTableIds = append(request.RouteTableIds, aws.String(routeTableId))
+		}
+		response, err := c.EC2().DescribeRouteTables(request)
+		if err != nil {
+			return nil, fmt.Errorf("error from DescribeRouteTables: %v", err)
+		}
+
+		for _, rt := range response.RouteTables {
+			for _, route := range rt.Routes {
+				if route.NatGatewayId != nil {
+					natGatewayIds.Insert(*route.NatGatewayId)
+				}
+			}
+		}
+	}
+
+	{
+		// We could do this to find if a NAT gateway is shared
+
+		//request := &ec2.DescribeRouteTablesInput{}
+		//request.Filters = append(request.Filters, awsup.NewEC2Filter("route.nat-gateway-id", natGatewayId))
+		//response, err := c.EC2().DescribeRouteTables(request)
+		//if err != nil {
+		//	return fmt.Errorf("error from DescribeRouteTables: %v", err)
+		//}
+		//
+		//for _, rt := range response.RouteTables {
+		//	routeTableId := aws.StringValue(rt.RouteTableId)
+		//}
+	}
+
+	var trackers []*ResourceTracker
+
+	{
+		request := &ec2.DescribeNatGatewaysInput{}
+		for natGatewayId := range natGatewayIds {
+			request.NatGatewayIds = append(request.NatGatewayIds, aws.String(natGatewayId))
+		}
+		response, err := c.EC2().DescribeNatGateways(request)
+		if err != nil {
+			return nil, fmt.Errorf("error from DescribeNatGateways: %v", err)
+		}
+
+		if response.NextToken != nil {
+			return nil, fmt.Errorf("NextToken set from DescribeNatGateways, but pagination not implemented")
+		}
+
+		for _, t := range response.NatGateways {
+			ngwTracker := &ResourceTracker{
+				Name:    aws.StringValue(t.NatGatewayId),
+				ID:      aws.StringValue(t.NatGatewayId),
+				Type:    TypeNatGateway,
+				deleter: DeleteNatGateway,
+			}
+			trackers = append(trackers, ngwTracker)
+
+			// If we're deleting the NatGateway, we should delete the ElasticIP also
```

Comment:
Another reference to here https://github.com/kubernetes/kops/issues/1368 just in case we get into this situation later where we have multiple NGW's to an EIP or something